### DISCUSSION TITLE
Wrong returned error type after rolling back an already committed transaction?

### DISCUSSION THREAD

**Discussion by LaCodon** [jackc/pgx#1990] [Q&A] [answered]
Hi everyone, I tried using `pgx` as a `database/sql` implementation and relied on the `tx.Rollback` function to return a `sql.ErrTxDone` (as defined here: https://pkg.go.dev/database/sql#pkg-variables) error in case the transaction has already been committed beforehand. However, the `pgx` transaction returns the custom `ErrTxClosed` error (https://github.com/jackc/pgx/blob/8f69e45a53a3954f6010eb9dd18c12a8105925ec/tx.go#L201).

Is there a reason why the error from the `database/sql` package is not used or not even wrapped at least? Should we change that?

**Comment by jackc** [comment, marked as answer]
It's not used in `pgx/tx.go` because that code has nothing to do with `database/sql`. That is the pgx native interface. As far as I know, `sql.ErrTxDone` should only be returned from `database/sql` itself.

**Reply by LaCodon** [reply]
Oh okay, I get it. I assumed, that `database/sql` just passes the driver's errors through to the user but it actually has its own logic for tracking the transaction state and returns `sql.ErrTxDone` based on that (for reference: https://cs.opensource.google/go/go/+/master:src/database/sql/sql.go;l=2309;drc=c23579f031ecd09bf37c644723b33736dffa8b92).

Many thanks for your fast response!
